rust/agama-dbus-server/src/network/model.rs

```diff
@@ -81,6 +90,52 @@ impl NetworkState {
         Ok(())
     }
 
+    /// Updates a controller connection with a new one.
+    ///
+    /// It uses the `id` to decide which connection to update.
+    ///
+    /// Additionally, it registers the connection to be removed when the changes are applied.
+    pub fn update_controller_connection(
+        &mut self,
+        mut conn: Connection,
+        settings: HashMap<String, ControllerSettings>,
+    ) -> Result<(), NetworkStateError> {
+        // let Some(old_conn) = self.get_connection_mut(conn.id()) else {
+        //     return Err(NetworkStateError::UnknownConnection(conn.id().to_string()));
+        // };
+        //
+        let mut new_ports = vec![];
+        let controller = conn.clone();
+
+        if let Connection::Bond(ref mut bond) = conn {
+            if let Some(ControllerSettings::Options(opts)) = settings.get("options") {
+                bond.bond.options = BondOptions::try_from(opts.clone()).unwrap()
+            }
+
+            if let Some(ControllerSettings::Ports(ports)) = settings.get("ports") {
+                for port in ports.iter() {
+                    new_ports.push(
+                        if let Some(new_port) =
+                            bond.bond.ports.iter().find(|c| c.interface() == port)
```

Comment:
Perhaps we could move this stuff to a `BondConfig` function. With a good name, it would be more expressive.

Reply:
Actually, you could follow an approach like this (I know, it does not compile, it is just an idea):

```rust
new_port = bond
  .find_port(port)
  .cloned()
  .unwrap_or_else(|| {
    // ... here the code to build the port
  });
```

In my opinion, it expresses what you want better. But perhaps it is just me.

About building the port, your code is perfectly fine. But perhaps we could try something like [the builder pattern](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html):

```rust
ConnectionBuilder::new(port.to_string(), DeviceType::Ethernet)
  .with_interface(port),
  .with_controller(controller.id())
  .build()
```